WIP: Python: wrap and patch using `requiredPythonModules` instead of `propagatedBuildInputs`

FRidh opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

Python programs look up their dependencies dynamically. To support this,
Python executables are wrapped and patched.

To find the dependencies to include when wrapping and patching, the
`wrapPythonPrograms` function would traverse the
`propagatedBuildInputs`. `propagatedBuildInputs` are dependencies that
need to be propagated during *build time*. While that is needed, we do
not want to have *all* propagated build inputs always as run-time
dependencies. Therefore, many derivations instead choose to set
`pythonPath`, which `wrapPythonPrograms` also considers.`pythonPath`,
however, is just an attribute that is not recursed into.

This commit implements `requiredPythonModules`, which was already set in
the `passthru` of Python packages for
`python.buildEnv`/`python.withPackages`. Now, it is also passed to the
builder, and `wrapPythonPrograms` will consider this environment
variable instead of `propagatedBuildInputs`. This change will thus
prevent unwanted bloat in wrappers. It also effectively replaces
`pythonPath`. Note we could have used `pythonPath` for this, but I want
to avoid that name because what we do is not exactly what `PYTHONPATH`
does.

Builders other than `buildPythonPackage` will typically need to pass
something like

    requiredPythonModules = with python3.pkgs; computeRequiredPythonModules [ sphinx ];

for run-time Python dependencies.

This change is a prerequisite to unifying how we deal with wrappers in
Nixpkgs. E.g., now we can look into a common method for dealing with
wrappers for Qt, Python and GApps without getting bloat from Python
packages.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).


### Tasks

- [x] Tested diffoscope. Closure size remained the same.
- [ ] change log stating backwards incompatible change
- [ ] update docs
- [x] replace `propagatedBuildInputs` treewide
- [x] replace `pythonPath` treewide

This is a backwards incompatible change. cc @jonringer as well as @adisbladis @DavHau @costrouc because of external tools.

Fixes https://github.com/NixOS/nixpkgs/issues/24128

FRidh committed:
pythonPackages: rename requiredPythonModules to computeRequiredPythonModules

requiredPythonModules is already an attribute in the (passthru of the)
builder.

FRidh committed:
WIP: Python: wrap and patch using `requiredPythonModules` instead of `propagatedBuildInputs`

Python programs look up their dependencies dynamically. To support this,
Python executables are wrapped and patched.

To find the dependencies to include when wrapping and patching, the
`wrapPythonPrograms` function would traverse the
`propagatedBuildInputs`. `propagatedBuildInputs` are dependencies that
need to be propagated during *build time*. While that is needed, we do
not want to have *all* propagated build inputs always as run-time
dependencies. Therefore, many derivations instead choose to set
`pythonPath`, which `wrapPythonPrograms` also considers.`pythonPath`,
however, is just an attribute that is not recursed into.

This commit implements `requiredPythonModules`, which was already set in
the `passthru` of Python packages for
`python.buildEnv`/`python.withPackages`. Now, it is also passed to the
builder, and `wrapPythonPrograms` will consider this environment
variable instead of `propagatedBuildInputs`. This change will thus
prevent unwanted bloat in wrappers. It also effectively replaces
`pythonPath`. Note we could have used `pythonPath` for this, but I want
to avoid that name because what we do is not exactly what `PYTHONPATH`
does.

Builders other than `buildPythonPackage` will typically need to pass
something like

requiredPythonModules = with python3.pkgs;
computeRequiredPythonModules [ sphinx ];

for run-time Python dependencies.

This change is a prerequisite to unifying how we deal with wrappers in
Nixpkgs. E.g., now we can look into a common method for dealing with
wrappers for Qt, Python and GApps without getting bloat from Python
packages.

TODO: update documentation

buildPython*: drop support for pythonPath

buildPythonPackage: add note regarding propagation

buildPython*: write requiredPythonModules to nix-support/

We would like to keep references to run-time dependencies. Before, this
was done with `propagatedBuildInputs` resulting in
`nix-support/propagated-build-inputs`.

This is also another step towards getting rid of `propagatedBuildInputs`.

FRidh committed:
treewide: buildPython*: use requiredPythonModules instead of propagatedBuildInputs

Python packages should now use `requiredPythonModules` to declare their
runtime dependencies.

Note in some cases there may have been good reasons to put a non-Python
package in `propagatedBuildInputs`. In that case, a manual change needs
to be made to put that package back into `propagatedBuildInputs`.

FRidh commented on doc/languages-frameworks/python.section.md:
```diff
@@ -899,7 +899,7 @@ following are specific to `buildPythonPackage`:
   `--install-option`. E.g., `pipInstallFlags=["--install-option='--cpp_implementation'"]`.
 * `pythonPath ? []`: List of packages to be added into `$PYTHONPATH`. Packages
-  in `pythonPath` are not propagated (contrary to `propagatedBuildInputs`).
+  in `pythonPath` are not propagated (contrary to `requiredPythonModules`).
```
need fixing

FRidh commented on doc/languages-frameworks/python.section.md:
```diff
@@ -920,7 +920,7 @@ because their behaviour is different:
   are added to `nativeBuildInputs` when `doCheck = true`. Items listed in
   `tests_require` go here.
-* `propagatedBuildInputs ? []`: Aside from propagating dependencies,
+* `requiredPythonModules ? []`: Aside from propagating dependencies,
```
need fixing

FRidh commented on pkgs/top-level/python-packages.nix:
```diff
@@ -79,7 +79,7 @@ let
         pythonModule = python;
         pythonPath = [ ]; # Deprecated, for compatibility.
-        requiredPythonModules = computeRequiredPythonModules drv.propagatedBuildInputs;
+        requiredPythonModules = computeRequiredPythonModules [ drv ] ++ drv.requiredPythonModules;
```
need parenthesis

FRidh committed:
treewide: replace pythonPath with requiredPythonModules

These are now effectively the same.

FRidh committed:
WIP treewide: replace pythonPath with requiredPythonModules (part 2)

These are the ones explicitly invoking wrapPythonProgramsIn with
pythonPath. Need to test this.

FRidh committed:
diffoscope: add non-Python dependencies to PATH

`wrapPython` will already extend PATH with the bin/ of each Python
dependency. We should do the same for non-Python dependencies.

FRidh committed:
kupfer: fix build

FRidh committed:
pythonPackages.pygobject3: cairo is not a Python dep

FRidh committed:
treewide: replace pythonPath for non-buildPython* packages

Replace pythonPath with requiredPythonModules for non-buildPython*
packages.

In this case, it is just a name of the attribute, and it does not matter
what it is because `wrapPythonProgramsIn` is invoked explicitly.

In a future commit wrapPython hook will be replaced by one that does not
require an explicit invocatoin of `wrapPythonProgramsIn`.

SuperSandro2000 commented:
``computeRequiredPythonModules`` this will be very hard to type correct on first try. I think it would be easier if the name would be shorter.

Also I personally have a little trouble reading and understanding very short lines which made reading the text a little bit hard.

DavHau commented:
Thanks for notifying me.
It would be nice if backwards incompatible changes like this would be discoverable, so that tools like mach-nix can operate on different versions of nixpkgs. Right now one could check for the existence of `python3.pkgs.computeRequiredPythonModules` to see if the new API needs to be used or not.
Though, it would be nicer if the version of nixpkgs' python-API would be versioned and that version was discoverable.

jonringer commented:
One consequence of this is that it will make backporting certain changes more difficult for people who are not aware of these changes.

It might be nice to do have some deprecation where, "Using `propagatedBuildInputs` in a `buildPython{Package,Application}` has been deprecated, please use requirePythonModules instead".

I have some locally packaged python packages that I use for work, and it would be a terrible user experience to try and debug why my previously working python packages aren't working.

FRidh commented:
Yes, backporting is more difficult, and out-of-tree code will break.

> It might be nice to do have some deprecation where, "Using propagatedBuildInputs in a buildPython{Package,Application} has been deprecated, please use requirePythonModules instead".

That's unfortunately not possible, because there are still legitimate cases for using `propagatedBuildInputs`. Now, typically we try to avoid propagation, but it is a feature of `mkDerivation` which I don't think we should break.

jonringer commented:
> That's unfortunately not possible, because there are still legitimate cases for using `propagatedBuildInputs`

I'm not saying breaking it, I'm just saying emit a `trace.warn` and continue the old behavior until people are aware of it. Maybe after the next release, then remove it. Not a lot of people have the "luxury" of watching nixpkgs super closely.

jonringer commented:
something like:

```

let
  pythonBuild = mkDerivation rec {
     ...
  }
in if lib.hasAttr "propagatedBuildInputs" attrs
  then
    lib.warn "Using propagatedBuildInputs to pass python packages has been deprecated in nixpkgs, please use requiredPythonModules instead. See https://github.com/NixOS/nixpkgs/pull/102613 for justification" pythonBuild
  else
    pythonBuild
```

jonringer commented:
The reality is that these paradigms and conventions have existed for many years, and changing those paradigms will be difficult for people. Especially people that may be using many channels, or manages the installation for many people as well.

FRidh commented:
Given Nixpkgs will be fixed directly, it will only create warnings for user packages, which is I think acceptable.

jonringer commented:
I think this is going to cause a lot of pain, especially with people who use the release channels.

> Note, however, this fixes only buildPython* usage, and not when using mkDerivation directly, unless a warning would also/instead be generated during build-time by wrapPython.

I think this is a lot less common than buildPython* usage.

jonringer commented:
> Given Nixpkgs will be fixed directly, it will only create warnings for user packages, which is I think acceptable. Also, it is possible to check whether `requiredPythonModules` is set, and only warn when not set.

There's also example and documentation that lives outside of the repo as well. for example, https://www.youtube.com/watch?v=jXd-hkP4xnU 

Please let me try to come up with a decent deprecation mechanism. People who do have out of tree packages will have a really painful time trying to debug why python packages don't seem to be available.

FRidh commented:
> Please let me try to come up with a decent deprecation mechanism. People who do have out of tree packages will have a really painful time trying to debug why python packages don't seem to be available.

Do that. I think what we just discussed is reasonable and doable already.

FRidh commented:
With b273cacf01dc3ecca2c915a79eedc09ee98fa0c5 we could have the Python setup hook, and the wrapper hook read those files as well. That way, we effectively implement our own propagation entirely and don't use `propagatedBuildInputs` anymore at all.

That reminds me again. The issue right now is that the wrappers won't be generated. That is the main problem regarding backwards compatibility. But in case of `buildPython*` we could assign `propagatedBuildInputs` to `requiredPythonModules` as well, so that issue is solved.

FRidh commented on pkgs/development/python-modules/pygobject/3.nix:
```diff
@@ -19,7 +19,8 @@ buildPythonPackage rec {
   buildInputs = [ glib gobject-introspection ]
                  ++ stdenv.lib.optionals stdenv.isDarwin [ which ncurses ];
-  requiredPythonModules = [ pycairo cairo ];
+  propagatedBuildInputs = [ cairo ];
```
@jtojnar should we propagate this dep you think? I prefer to avoid it, but did it now just to be safe

jtojnar replied:
I do not see why it should be propagated. Gdk overrides mention Cairo typelib but that is only for old versions of Gtk and the typelib is in `gobject-introspection` package anyway. The only reason it would be propagated is for compilation but it is not listed in pkg-config file so I doubt it is the case.

It was propagated in https://github.com/NixOS/nixpkgs/pull/5893 without any explanation.

jonringer commented:
I think I came up with something useful:
```
[10:35:00] jon@jon-desktop /home/jon/projects/nixpkgs (requiredPythonModules)
$ nix-build -A python3Packages.hypothesis
trace: warning: Using propagatedBuildInputs for python modules has been deprecated
trace: warning: Using propagatedBuildInputs for python modules has been deprecated
these derivations will be built:
  /nix/store/zi6xgblclc0z4cb7211mp8mql340sp4w-python3.8-hypothesis-5.30.0.drv
waiting for locks or build slots...
^Cerror: interrupted by the user

# changed propagatedBuildInputs -> requiredPythonModules

[10:36:22] jon@jon-desktop /home/jon/projects/nixpkgs (requiredPythonModules)
$ nix-build -A python3Packages.hypothesis
these derivations will be built:
  /nix/store/zi6xgblclc0z4cb7211mp8mql340sp4w-python3.8-hypothesis-5.30.0.drv
```

it still evaluates to the same derivation

```
$ git diff -- pkgs/development/interpreters/python/
diff --git a/pkgs/development/interpreters/python/mk-python-derivation.nix b/pkgs/development/interpreters/python/mk-python-derivation.nix
index 631aadc1a40..d645a70014c 100644
--- a/pkgs/development/interpreters/python/mk-python-derivation.nix
+++ b/pkgs/development/interpreters/python/mk-python-derivation.nix
@@ -105,7 +105,16 @@ else
 let
   inherit (python) stdenv;

-  requiredPythonModules_ = computeRequiredPythonModules requiredPythonModules;
+  #
+  partitionedPropagatedInputs = lib.lists.partition (x: lib.hasAttr "pythonModule" x) propagatedBuildInputs;
+  propagatedPythonInputs = if lib.length partitionedPropagatedInputs.right > 0
+    then
+      lib.warn "Using propagatedBuildInputs for python modules has been deprecated" partitionedPropagatedInputs.right
+    else
+      [ ];
+  propagatedDerivationInputs = partitionedPropagatedInputs.wrong;
+
+  requiredPythonModules_ = computeRequiredPythonModules (requiredPythonModules ++ propagatedPythonInputs);

   self = toPythonModule (stdenv.mkDerivation ((builtins.removeAttrs attrs [
     "disabled" "checkPhase" "checkInputs" "doCheck" "doInstallCheck" "dontWrapPythonPrograms" "catchConflicts" "format" "requiredPythonModules"
@@ -157,7 +166,7 @@ let
     # Probably the solution is to keep using propagation in case packages,
     # and convert `buildPythonApplication` to first build the package, and then an environment
     # using `python.buildEnv` https://github.com/NixOS/nixpkgs/pull/16672.
-    propagatedBuildInputs = propagatedBuildInputs ++ requiredPythonModules_;
+    propagatedBuildInputs = propagatedDerivationInputs ++ requiredPythonModules_;

     # requiredPythonModules is also set
     requiredPythonModules = requiredPythonModules_;
```

jonringer commented:
I think the message should be:

> Using propagatedBuildInputs for python modules has been deprecated, please use requiredPythonModules instead. See https://github.com/NixOS/nixpkgs/pull/102613 for justification of change.

FRidh committed:
buildPythonPackage: warn if propagatedBuildInputs is used

Now that Python packages should use `requiredPythonModules` instead of
`propagatedBuildInputs`, we could actually check it during evaluation time.

This commit adds a warning in case Python modules are propagated. Now, in
principle it is fine to propagate modules, but there is not really any
good reason for it.

Note application propagation is not affected by this because in that case
`drv.pythonModule` is `false` or missing.

FRidh committed:
claws-mail: fix eval

FRidh commented:
Introduces a warning message that also suggests how the code should look.

Is there a better name than `requiredPythonModules`? First thing I would ask when I see it is "required for what"?

Maybe also worth adding a note that it may be tempting to use the file in `$out/nix-support/` to get rid of `requiredPythonModules` altogether, however, at that point you lose the separation between build-time and run-time dependencies, and end up with the current situation.

FRidh committed:
fixup: filter null

FRidh committed:
claws-mail-gtk3: fix eval

jonringer commented:
> Is there a better name than `requiredPythonModules`? First thing I would ask when I see it is "required for what"?

I think `required` is implicit in it being defined at all. `pythonModules` would suffice, if they weren't required, then they wouldn't be listed.

Also, I would prefer calling it `pythonPackages`, as some packages such as `setuptools` can export many top-level modules; and our level of abstraction is a package, not individual modules.

jonringer commented:
I'm a fan of the deprecation message.

jonringer commented:
would we lose any functionality by changing:
```
propagatedBuildInputs = propagatedDerivationInputs ++ requiredPythonModules_;
```
to
```
propagatedBuildInputs = propagatedDerivationInputs;
```
seems weird to filter out the python packages, just to reintroduce them again.

FRidh committed:
meson: don't remove propagated-build-inputs

BuildPython* no longer propagates dependencies that way.

FRidh committed:
gtk-doc: no need to clear Python from propagated-build-inputs anymore

since its no longer propagated.

FRidh committed:
Python: get rid of propagation altogether

FRidh committed:
treewide: further convert propagatedBuildInputs to requiredPythonModules

FRidh commented:
No more `propagatedBuildInputs` at all now! The part of propagation that is needed, is now done explicitly by Python's setup hook for build-time `PYTHONPATH`, and by `wrapPython` for run-time.

> I think required is implicit in it being defined at all. pythonModules would suffice, if they weren't required, then they wouldn't be listed.

> Also, I would prefer calling it pythonPackages, as some packages such as setuptools can export many top-level modules; and our level of abstraction is a package, not individual modules.

In the Python world the following names exist:

|Project| setuptools | flit | poetry | pep 621 draft | 
|---------|----------------|----|----------|--------------------|
|name | install_requires | requires | dependencies | dependencies | 

https://www.python.org/dev/peps/pep-0621/#dependencies-optional-dependencies

We need to explicitly add Python to it. I find the names in the table bad because, except for setuptools, they don't define what type of dependencies they are. Thus, I am in favor of something like `pythonInstallDeps` or `pythonRuntimeDeps`.

FRidh committed:
fixup get rid of propagation altogether

FRidh committed:
python.pkgs.matplotlib: fix expression

FRidh committed:
python2.pkgs.pysqlite: fix expression

FRidh commented:
Hydra is evaluating this PR at https://hydra.nixos.org/jobset/nixpkgs/python-unstable. If that looks OK, we need to make a dedicated job evaluating the whole of Nixpkgs. As soon as that is OK, we rebase it onto the latest staging-next and have it in staging-next the cycle after.

FRidh committed:
gtklick: fix expression

FRidh committed:
gtklick: fix expression

FRidh committed:
mirage-im: fix evaluation

FRidh committed:
pythonPackages.cairocffi: fix expression

FRidh committed:
yle-dl: fix expression

FRidh committed:
pythonPackages: revert to python as parameter

That way we don't break poetry2nix.

FRidh committed:
pythonPackages.pendulum: remove unused parameter poetry

jonringer commented:
> We need to explicitly add Python to it. I find the names in the table bad because, except for setuptools, they don't define what type of dependencies they are. Thus, I am in favor of something like `pythonInstallDeps` or `pythonRuntimeDeps` to distinguish between build-time dependencies and test-time dependencies.

Hmm, I think the terminology might be confusing as we have mismatched terms (nix vs python).

Even in the `pythonInstallDeps` or `pythonRuntimeDeps`, I thought you were relating these to `setup_requires` and `install_requires`, and didn't release you  were talking about build and test-time dependencies.

I would prefer the term `Inputs` over `Deps`, to keep the number of "new but similar terms" to a minimum.

Personally, I would do the following mapping, keeping a preference for Nix related terminology:

| Python wrapper | related nix term | example |
| --- | --- | --- |
| pythonNativeInputs | nativeBuildInputs | cython |
| pythonRuntimeInputs | (current) propagatedBuildInputs | requests |
| pythonCheckInputs | checkInputs | pytest |

And then, in the documentation just have a chart of how these terms relate.

FRidh commented:
> Even in the pythonInstallDeps or pythonRuntimeDeps, I thought you were relating these to setup_requires and install_requires, and didn't release you were talking about build and test-time dependencies.

Yes, that was a *very* bad sentence of mine. Updated it.

So actually, there is no need whatsoever for specific Python inputs. We actually handle everything with the normal inputs to build things. The only reason for having `requiredPythonModules` is to keep a package working during runtime (by wrapping and patching). In `buildPython*` the `requiredPythonModules` get added to `buildInputs` just for convenience, otherwise you would have to list them twice. When not using `buildPython*` you need to set both `buildInputs` and the new `requiredPythonModules`. This "convenience" logic could be removed from `buildPython*`, but I don't think anyone gets happy with that either (it would break every package).

There are other builders that use `Deps`. My motivation to choose `Deps` is to exactly differentiate with the actual `Inputs` `mkDerivation` needs. But to be clear, this part I don't care so much about.

FRidh committed:
pythonPackages.jsbeautifier: its in the package set, so buildPythonPackage

FRidh committed:
fixup diffoscope

FRidh committed:
fixup buildPython*: enhance warning

FRidh committed:
python.pkgs.binwalk: fix expression

FRidh committed:
pythonPackages.hidapi: fix expression

DavHau commented:
> In buildPython* the requiredPythonModules get added to buildInputs just for convenience, otherwise you would have to list them twice

Why would one need to list them twice? Python runtime deps are not required during build time and therefore should not be put in any buildInputs I'd argue. I think it would be best if python runtime deps would not have any effect on a package's build at all. This would remove the necessity to rebuild packages if their runtime deps change, which is a big deal for usability in my opinion.

The other way round, python build time deps should only end up in buildInputs and not be kept as a runtime dep.

If a python dep is needed as a build time AND runtime dep, then of course you would need to add them twice. But I guess this case is quite rare.

FRidh commented:
> Python runtime deps are not required during build time and therefore should not be put in any buildInputs I'd argue. 

Yes, they are. During our build-time we also install Python packages, and for that it is needed. Unless we tell pip to ignore testing whether all install requirements are full-filled. Its important to be clear which build-time we're talking about, from Nix point of view, or from Python point of view.

What you are suggesting is separating wheel-building from wheel-installing. That would be a larger change. This gets us closer to that as well. During wheel-building, the packages that are used for building should be in `nativeBuildInputs` and those that we link against (e.g. when compiling numpy) should be in `buildInputs`.

FRidh committed:
libvirt: use python3 as nativeBuildInput

FRidh committed:
pythonPackages.pyscard: wrap non-Python executables

FRidh committed:
pythonPackages.trezor: fix buildInputs

FRidh committed:
pythonPackages.gst-python: fix buildInputs

FRidh committed:
pythonPackages.pyparted: fix buildInputs

FRidh commented:
I was worried about evaluation performance regressions, but I could not find any degradation with `NIX_SHOW_STATS=1`. Quite the contrary, with every package I checked every stat dropped in value.

FRidh commented:
About `requiredPythonModules`. As it is, it is only relevant for storing the runtime dependencies, such as for wrapping. 

During the build, `PYTHONPATH` is configured using Python's setup hook which considers the `hostOffset`. The pro of this is that it is a simple solution requiring only buildInputs to be set. The con is that it does not distinguish between build and check inputs.

The alternative is to specify specifically which Python inputs should be added when (as suggested by @jonringer). That would mean having hooks running at different phases to setup `PYTHONPATH` *as well as* `PATH`. In my opinion we would trespassing into `stdenv` space then. I would prefer to do this with multiple derivations instead (https://github.com/NixOS/nixpkgs/issues/118452#issuecomment-813006581).

FRidh commented:
> requiredPythonModules = with python3.pkgs; computeRequiredPythonModules [ sphinx ];

This could be made unnecessary if Python's setup hook would check the file in `nix-support` and recurse in those folders.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-shell-buildinputs-ordering-issue/12885/4

LunNova commented:
What's still left to do here?

jonringer commented:
it's already implemented for the most part. The major thing is to make it so that people stop using propagatedBuildInputs by default.

LunNova commented:
I think this will make `overrideAttrs` not work for `requiredPythonModules`

https://github.com/NixOS/nixpkgs/pull/102613/commits/6f845f7622dab936ff76a7c340134a54decd724f#diff-c263b18f23de7c1f00ed846783c3b977718aae8cdfdd89fc82adcf365d1d8bebR108-R163

If we `overrideAttrs` for `requiredPythonModules`, the old copy in `requiredPythonModules_` will still be used.

If we wait for #119942 to go in we could refer to `finalAttrs.requiredPythonModules` instead.

yajo commented:
Hello! I think this is a great improvement.

Very often I see hundreds of python packages rebuilt without need because we don't have such patch.

It's sad that it's become so stale. Now it has more than 3000 lines of conflict.

I suggest a simpler path: just add a valid deprecation path that keeps old derivations working as usual, but supports this new way of doing it. This way, python packages can evolve gradually to adopt this new system. Also the PR won't have 3000 lines and it will be easier to review and merge. WDYT?

yajo commented:
#271597 implemented this
